### PR TITLE
feat: Handle nested italics in bold text (issue 2324)

### PR DESCRIPTION
**Pull request by btertoolen**
Fixes issue https://github.com/AppFlowy-IO/AppFlowy/issues/2324

I ended up creating a single transaction containing all the operations needed to handle the nested italics, as calling editorState.apply() multiple times in the handler function was giving me issues

### CODE REVIEW

File: lib/src/service/internal_key_event_handlers/markdown_syntax_to_styled_text.dart [hunk 1]
```diff
@@ -365,21 +365,93 @@ ShortcutEventHandler doubleAsteriskToBoldHandler = (editorState, event) {
...
+      }
+    }
 
-  editorState.apply(transaction);
+// Remove all underscore characters in the string
+// If we have an uneven number of underscore characters we keep the last character
+// as it did not form an italics segment
+    if (nestedItalicsIndexes.length % 2 == 1) {
+      nestedItalicsIndexes.removeLast();
+    }
```
Comment (Xazin) [inline lib/src/service/internal_key_event_handlers/markdown_syntax_to_styled_text.dart:424]: Can you add a test for this scenario? And maybe a linebreak between this if statement and the next for-loop.

File: lib/src/service/internal_key_event_handlers/markdown_syntax_to_styled_text.dart [hunk 2]
```diff
@@ -365,21 +365,93 @@ ShortcutEventHandler doubleAsteriskToBoldHandler = (editorState, event) {
...
+    for (var i = 0; i < nestedItalicsIndexes.length; i++) {
+      // Since we are deleting multiple items in a single transaction, everytime
+      // we delete a character the next deletion's index must be decremented by
+      // the amount of characters we have removed, to compensate for the shifted
+      // indexes after the deletion
+      var transactionDelete = editorState.transaction
+        ..deleteText(textNode, nestedItalicsIndexes[i] - i, 1);
+      transaction.operations.addAll(transactionDelete.operations);
```
Comment (Xazin) [inline lib/src/service/internal_key_event_handlers/markdown_syntax_to_styled_text.dart:432]: Can be final, also you don't need to use `addAll`.

You can just do `transaction.deleteText(textNode, nestedItalicsIndices[i] - 1, 1);`.

File: lib/src/service/internal_key_event_handlers/markdown_syntax_to_styled_text.dart [hunk 3]
```diff
@@ -365,21 +365,93 @@ ShortcutEventHandler doubleAsteriskToBoldHandler = (editorState, event) {
...
+
+    // Remove the asterixes from the text
+    var transactionDelete = editorState.transaction
+      ..deleteText(textNode, thirdToLastAsteriskIndex, 2)
+      ..deleteText(
+        textNode,
+        lastAsteriskIndex - 2 - nestedItalicsIndexes.length,
+        1,
+      );
+    transaction.operations.addAll(transactionDelete.operations);
```
Comment (Xazin) [inline lib/src/service/internal_key_event_handlers/markdown_syntax_to_styled_text.dart:443]: Same here, you could just do
```suggestion
    transaction
      ..deleteText(textNode, thirdToLastAsteriskIndex, 2)
      ..deleteText(
        textNode,
        lastAsteriskIndex - 2 - nestedItalicsIndexes.length,
        1,
      );
```

File: test/service/internal_key_event_handlers/markdown_syntax_to_styled_text_test.dart [hunk 4]
```diff
@@ -379,6 +379,210 @@ void main() async {
         expect(textNode.toPlainText(), text);
       }),
     );
+
+    testWidgets(
+      '**bold and _nested_ italics**',
+      ((widgetTester) async {
```
Comment (Xazin) [inline test/service/internal_key_event_handlers/markdown_syntax_to_styled_text_test.dart:385]: One paranthesis too many

File: test/service/internal_key_event_handlers/markdown_syntax_to_styled_text_test.dart [hunk 5]
```diff
@@ -379,6 +379,210 @@ void main() async {
...
+        expect(allTextBold, true);
+        expect(segmentItalic, true);
+        expect(lastSegmentItalic, false);
+        expect(plainText, plainResult);
+      }),
+    );
+
+    testWidgets(
+      'regular then **bold and _nested_ italics**',
+      ((widgetTester) async {
```
Comment (Xazin) [inline test/service/internal_key_event_handlers/markdown_syntax_to_styled_text_test.dart:446]: Same here, one paranthesis too many.

File: test/service/internal_key_event_handlers/markdown_syntax_to_styled_text_test.dart [hunk 6]
```diff
@@ -379,6 +379,210 @@ void main() async {
...
+        expect(textInAsterixesIsBold, true);
+        expect(segmentItalic, true);
+        expect(lastSegmentItalic, false);
+        expect(plainText, plainResult);
+      }),
+    );
+
+    testWidgets(
+      '**bold and _double_ _nested_ italics**',
+      ((widgetTester) async {
```
Comment (Xazin) [inline test/service/internal_key_event_handlers/markdown_syntax_to_styled_text_test.dart:514]: Same here, one paranthesis too many.

File: lib/src/service/internal_key_event_handlers/markdown_syntax_to_styled_text.dart [hunk 7]
```diff
@@ -365,21 +365,93 @@ ShortcutEventHandler doubleAsteriskToBoldHandler = (editorState, event) {
...
+  for (var i = 0; i < text.length; i++) {
+    if (text[i] == '_') {
+      nestedItalicsIndexes.add(i);
+    }
+  }
+  if (nestedItalicsIndexes.length > 1) {
+    textContainsNestedItalics = true;
+  }
+
+  if (!textContainsNestedItalics) {
```
[the remark this answers is not in the document]
Comment (btertoolen) [reply lib/src/service/internal_key_event_handlers/markdown_syntax_to_styled_text.dart:379]: Wouldn't this give a false positive if a string only contains a single underscore character?

File: lib/src/service/internal_key_event_handlers/markdown_syntax_to_styled_text.dart (hunk 2 above)
Comment (btertoolen) [reply lib/src/service/internal_key_event_handlers/markdown_syntax_to_styled_text.dart:432]: I will update it

File: lib/src/service/internal_key_event_handlers/markdown_syntax_to_styled_text.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (Xazin) [reply lib/src/service/internal_key_event_handlers/markdown_syntax_to_styled_text.dart]: Fair enough, then just change it to `if (nestedItalicsIndices.length > 1)`
The rest of the change is still an improvement though, you don't need the `textContainsNestedItalics` boolean.

[the remark this answers is not in the document]
Comment (btertoolen) [reply lib/src/service/internal_key_event_handlers/markdown_syntax_to_styled_text.dart]: Sure, I updated it in the latest commit

**Comment by LucasXu0** [comment]
Hey, @btertoolen. The tests failed.

```
   info • The value of the local variable 'textContainsNestedItalics' isn't used • lib/src/service/internal_key_event_handlers/markdown_syntax_to_styled_text.dart:368:8 • unused_local_variable
```

**Comment by btertoolen** [comment]
@LucasXu0 Should be fixed now

**Pull request merged**